contracts/UniswapWrapper.sol

```diff
@@ -0,0 +1,176 @@
+    }
+
+    /**
+    * @dev It will send the received ETH to Uniswap to get bonded tokens,
+    *      and activate the converted tokens in the jurors registry instance of the Aragon Court.
+    * @param _minTokens Minimum amount of bonded tokens obtained in Uniswap
+    * @param _deadline Transaction deadline for Uniswap
+    * @param _activate Signal activation of tokens in the registry
+    */
+    function contributeEth(uint256 _minTokens, uint256 _deadline, bool _activate) external payable {
+        require(msg.value > 0, ERROR_ZERO_AMOUNT);
+
+        // get the Uniswap exchange for the bonded token
+        address payable uniswapExchangeAddress = uniswapFactory.getExchange(address(bondedToken));
+        require(uniswapExchangeAddress != address(0), ERROR_UNISWAP_UNAVAILABLE);
+        IUniswapExchange uniswapExchange = IUniswapExchange(uniswapExchangeAddress);
+
+        // swap tokens
+        uint256 bondedTokenAmount = uniswapExchange.ethToTokenSwapInput.value(msg.value)(_minTokens, _deadline);
+
+        // stake and activate in the registry
+        _stakeAndActivate(msg.sender, bondedTokenAmount, _activate);
+
+        // make sure there's no ETH left
+        uint256 ethBalance = address(this).balance;
+        if (ethBalance > 0) {
+            // solium-disable security/no-call-value
+            (bool result,) = msg.sender.call.value(ethBalance)("");
+            require(result, ERROR_ETH_REFUND);
+        }
+    }
+
+    function _contributeExternalToken(
+        address _from,
+        uint256 _amount,
+        address _token,
+        uint256 _minTokens,
+        uint256 _minEth,
+        uint256 _deadline,
+        bool _activate
+    )
+        internal
+    {
+        require(_amount > 0, ERROR_ZERO_AMOUNT);
+
+        // move tokens to this contract
+        ERC20 token = ERC20(_token);
+        require(token.safeTransferFrom(_from, address(this), _amount), ERROR_TOKEN_TRANSFER_FAILED);
+
+        // get the Uniswap exchange for the external token
+        address payable uniswapExchangeAddress = uniswapFactory.getExchange(_token);
+        require(uniswapExchangeAddress != address(0), ERROR_UNISWAP_UNAVAILABLE);
+        IUniswapExchange uniswapExchange = IUniswapExchange(uniswapExchangeAddress);
+
+        require(token.safeApprove(address(uniswapExchange), _amount), ERROR_TOKEN_APPROVAL_FAILED);
+
+        // swap tokens
+        uint256 bondedTokenAmount = uniswapExchange.tokenToTokenSwapInput(_amount, _minTokens, _minEth, _deadline, address(bondedToken));
+
+        // stake and activate in the registry
+        _stakeAndActivate(_from, bondedTokenAmount, _activate);
+
+        // refund leftovers if any
+        _refund(token);
+        _refund(bondedToken);
+    }
+
+    function _stakeAndActivate(address _from, uint256 _amount, bool _activate) internal {
+        // activate in registry
+        bondedToken.approve(address(registry), _amount);
+        bytes memory data;
+        if (_activate) {
```

Comment:
If activate is false, we should not stake and transfer the tokens to the wallet as we changed in the Phase 1 contract

Reply:
Yes, but what would be the point of using this wrapper to buy tokens for your own external wallet when you can just use Uniswap?
I can see why you would stake but not activate though: you plan to become a juror, you see the bonding curve has a good price and it looks ascending, but you don't want to become a juror right now.
If staking without activating is not an option, then I would remove the `activate` parameter at all and assume that this wrapper always stakes and activates.

Reply:
Fair enough! Was fixed in the Phase 1 mindset ;)

Reply:
I get it, it happened the opposite to me withe the previous one ;)